File: bindings/node/examples/typed.ts
import { rmSync } from 'node:fs'
import path from 'node:path'

import { SombraDB } from '../typed'
import type { GraphSchema } from '../typed'

const DB_PATH = './typed-example.db'

interface MyGraphSchema extends GraphSchema {
  nodes: {
    Person: { name: string; age: number }
    Company: { name: string; employees: number }
    City: { name: string; state: string }
    Pet: { name: string; species: string }
  }
  edges: {
    KNOWS: {
      from: 'Person'
      to: 'Person'
      properties: { since: number }
    }
    WORKS_AT: {
      from: 'Person'
      to: 'Company'
      properties: { role: string }
    }
    LIVES_IN: {
      from: 'Person'
      to: 'City'
      properties: Record<string, never>
    }
    OWNS: {
      from: 'Person'
      to: 'Pet'
      properties: { since: number }
    }
    PARENT_OF: {
      from: 'Person'
      to: 'Person'
      properties: { since: number }
    }
    MARRIED_TO: {
      from: 'Person'
      to: 'Person'
      properties: { since: number }
    }
  }
}

function resetDb(file: string): void {
  try {
    rmSync(file)
  } catch {
    /* ignore */
  }
}

async function run(): Promise<void> {
  resetDb(DB_PATH)
  const schema: MyGraphSchema = {
    nodes: {
      Person: { properties: { name: '', age: 0 } },
      Company: { properties: { name: '', employees: 0 } },
      City: { properties: { name: '', state: '' } },
      Pet: { properties: { name: '', species: '' } },
    },
    edges: {
      KNOWS: { from: 'Person', to: 'Person', properties: { since: 0 } },
      WORKS_AT: { from: 'Person', to: 'Company', properties: { role: '' } },
      LIVES_IN: { from: 'Person', to: 'City', properties: {} },
      OWNS: { from: 'Person', to: 'Pet', properties: { since: 0 } },
      PARENT_OF: { from: 'Person', to: 'Person', properties: { since: 0 } },
      MARRIED_TO: { from: 'Person', to: 'Person', properties: { since: 0 } },
    },
  }

  const db = new SombraDB<MyGraphSchema>(DB_PATH, { schema })
  console.log('=== Type-Safe SombraDB Demo ===\n')

  const fabian = db.addNode('Person', { name: 'Fabian', age: 32 })
  const michelle = db.addNode('Person', { name: 'Michelle', age: 33 })
  const levi = db.addNode('Person', { name: 'Levi', age: 4 })
  const sarah = db.addNode('Person', { name: 'Sarah', age: 29 })
  const daniel = db.addNode('Person', { name: 'Daniel', age: 31 })
  const carlos = db.addNode('Person', { name: 'Carlos', age: 28 })
  const auroraTech = db.addNode('Company', { name: 'AuroraTech', employees: 250 })
  const omniCorp = db.addNode('Company', { name: 'OmniCorp', employees: 1200 })
  const austin = db.addNode('City', { name: 'Austin', state: 'TX' })
  const mochi = db.addNode('Pet', { name: 'Mochi', species: 'Dog' })

  console.log('✅ Created nodes with full type safety!\n')

  db.addEdge(fabian, michelle, 'KNOWS', { since: 2015 })
  db.addEdge(fabian, levi, 'PARENT_OF', { since: 2021 })
  db.addEdge(michelle, levi, 'PARENT_OF', { since: 2021 })
  db.addEdge(fabian, michelle, 'MARRIED_TO', { since: 2018 })
  db.addEdge(fabian, sarah, 'KNOWS', { since: 2022 })
  db.addEdge(sarah, daniel, 'KNOWS', { since: 2023 })
  db.addEdge(carlos, fabian, 'KNOWS', { since: 2024 })
  db.addEdge(fabian, auroraTech, 'WORKS_AT', { role: 'Staff Software Engineer' })
  db.addEdge(michelle, auroraTech, 'WORKS_AT', { role: 'Product Manager' })
  db.addEdge(daniel, omniCorp, 'WORKS_AT', { role: 'Data Engineer' })
  db.addEdge(fabian, austin, 'LIVES_IN', {})
  db.addEdge(michelle, austin, 'LIVES_IN', {})
  db.addEdge(carlos, austin, 'LIVES_IN', {})
  db.addEdge(fabian, mochi, 'OWNS', { since: 2020 })

  console.log('✅ Created edges with full type safety!\n')

  console.log('1. Find company by name (type-safe property query):')
  const auroraId = db.findNodeByProperty('Company', 'name', 'AuroraTech')
  if (auroraId) {
    const auroraNode = db.getNode(auroraId, 'Company')
    console.log(
      `   Found: ${auroraNode?.properties.name} with ${auroraNode?.properties.employees} employees\n`,
    )
  }

  console.log('2. Get all employees at AuroraTech (type-safe edge traversal):')
  const employeeIds = db.getIncomingNeighbors(auroraTech, 'WORKS_AT')
  console.log(`   Found ${employeeIds.length} employees:`)
  for (const empId of employeeIds) {
    const emp = db.getNode(empId, 'Person')
    console.log(`   - ${emp?.properties.name} (age ${emp?.properties.age})`)
  }
  console.log()

  console.log('3. Using type-safe query builder:')
  const result = db
    .query()
    .startFromLabel('Company')
    .traverse(['WORKS_AT'], 'in', 1)
    .getIds()
  console.log(`   Found ${result.nodeIds.length} total employees across all companies\n`)

  console.log('4. Find all people in Austin:')
  const austinResidents = db.getIncomingNeighbors(austin, 'LIVES_IN')
  console.log(`   ${austinResidents.length} people live in Austin:`)
  for (const personId of austinResidents) {
    const person = db.getNode(personId, 'Person')
    console.log(`   - ${person?.properties.name}`)
  }
  console.log()

  console.log('5. Analytics with type-safe labels:')
  console.log(`   Total people: ${db.countNodesWithLabel('Person')}`)
  console.log(`   Total companies: ${db.countNodesWithLabel('Company')}`)
  console.log(`   Total WORKS_AT relationships: ${db.countEdgesWithType('WORKS_AT')}`)
  console.log(`   Total KNOWS relationships: ${db.countEdgesWithType('KNOWS')}`)
  console.log()

  console.log('6. BFS traversal from Fabian (depth 2):')
  const bfsResults = db.bfsTraversal(fabian, 2)
  console.log(`   Reached ${bfsResults.length} nodes:`)
  for (const { nodeId, depth } of bfsResults.slice(0, 5)) {
    const node = db.getNode(nodeId, 'Person')
    console.log(`   - ${node?.properties.name ?? 'N/A'} (depth: ${depth})`)
  }
  console.log()

  console.log('✅ All type-safe operations completed successfully!\n')
  console.log('=== Type Safety Benefits ===')
  console.log('- Autocomplete for all labels and edge types')
  console.log('- Property types enforced at compile time')
  console.log('- No runtime casting required')
  console.log('- Backed by the same rust core and traversal primitives')
}

if (path.basename(process.argv[1] ?? '') === path.basename(__filename)) {
  run().catch((err) => {
    console.error(err)
    process.exitCode = 1
  })
}

File: bindings/python/examples/typed.py
"""Typed SombraDB facade demo for Python users."""

from __future__ import annotations

from pathlib import Path
from typing import TypedDict

from typing_extensions import Literal

from sombra.typed import NodeSchema, SombraDB, TypedGraphSchema

DB_PATH = Path(__file__).with_name("typed-example.db")


class PersonProps(TypedDict):
    name: str
    age: int


class CompanyProps(TypedDict):
    name: str
    employees: int


class CityProps(TypedDict):
    name: str
    state: str


class WorksAtProps(TypedDict):
    role: str


class LivesInProps(TypedDict, total=False):
    pass


class PersonNode(NodeSchema):
    properties: PersonProps


class CompanyNode(NodeSchema):
    properties: CompanyProps


class CityNode(NodeSchema):
    properties: CityProps


class GraphNodes(TypedDict):
    Person: PersonNode
    Company: CompanyNode
    City: CityNode


WorksAtEdge = TypedDict(
    "WorksAtEdge",
    {
        "from": Literal["Person"],
        "to": Literal["Company"],
        "properties": WorksAtProps,
    },
)

LivesInEdge = TypedDict(
    "LivesInEdge",
    {
        "from": Literal["Person"],
        "to": Literal["City"],
        "properties": LivesInProps,
    },
)


class GraphEdges(TypedDict):
    WORKS_AT: WorksAtEdge
    LIVES_IN: LivesInEdge


class DemoGraphSchema(TypedGraphSchema):
    nodes: GraphNodes
    edges: GraphEdges


def reset_db(path: Path) -> None:
    try:
        path.unlink()
    except FileNotFoundError:
        pass


def build_schema() -> DemoGraphSchema:
    return {
        "nodes": {
            "Person": {"properties": {"name": "", "age": 0}},
            "Company": {"properties": {"name": "", "employees": 0}},
            "City": {"properties": {"name": "", "state": ""}},
        },
        "edges": {
            "WORKS_AT": {
                "from": "Person",
                "to": "Company",
                "properties": {"role": ""},
            },
            "LIVES_IN": {"from": "Person", "to": "City", "properties": {}},
        },
    }


def main() -> None:
    reset_db(DB_PATH)
    schema = build_schema()
    db = SombraDB(str(DB_PATH), schema=schema)

    print("=== Typed SombraDB Demo ===\n")

    fabian = db.add_node("Person", {"name": "Fabian", "age": 32})
    michelle = db.add_node("Person", {"name": "Michelle", "age": 33})
    aurora = db.add_node("Company", {"name": "AuroraTech", "employees": 250})
    austin = db.add_node("City", {"name": "Austin", "state": "TX"})

    db.add_edge(fabian, aurora, "WORKS_AT", {"role": "Staff Software Engineer"})
    db.add_edge(michelle, aurora, "WORKS_AT", {"role": "Product Manager"})
    db.add_edge(fabian, austin, "LIVES_IN", {})
    db.add_edge(michelle, austin, "LIVES_IN", {})

    print("1. Find a company by typed property lookup:")
    aurora_id = db.find_node_by_property("Company", "name", "AuroraTech")
    if aurora_id:
        record = db.get_node(aurora_id, "Company")
        if record:
            props = record["properties"]
            print(f"   Found {props['name']} with {props['employees']} employees\n")

    print("2. Collect employees via typed edges:")
    employees = db.get_incoming_neighbors(aurora, "WORKS_AT")
    for idx, emp_id in enumerate(employees, start=1):
        person = db.get_node(emp_id, "Person")
        if person:
            props = person["properties"]
            print(f"   {idx}. {props['name']} (age {props['age']})")
    print()

    print("3. Run the typed BFS helper:")
    traversal = db.query().start_from_label("Person").traverse(["LIVES_IN"], "out", 1).get_ids()
    print(f"   Reached {len(traversal['node_ids'])} unique cities\n")

    print("Done! Inspect typed-example.db to explore the data.")


if __name__ == "__main__":
    main()

